Add nilu air quality sensor platform

hfurubotten opened the pull request:
## Description:

The `nilu_air_quality` platform shows measurements of current air quality from NILU (Norsk Institutt for luftforskning/Norwegian Institute for Air Research) sensor stations within Norway. Makes data from the open api at luftkvalitet.info and nilu.no available in Home Assistant.

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#7325

## Example entry for `configuration.yaml` (if applicable):
```yaml
sensor:
  - platform: nilu_air_quality
    area: Brumunddal
    monitored_conditions:
      - index
      - 'NO'
      - NO2
      - NOx
      - PM1
      - PM2.5
      - PM10
      - O3
      - SO2
      - CO
  - platform: nilu_air_quality
    show_on_map: True
    monitored_conditions:
      - index
  - platform: nilu_air_quality
    station: Zeppelinfjellet
    show_on_map: True
    monitored_conditions:
      - O3
  - platform: nilu_air_quality
    show_on_map: True
    name: Pollution Bergen
    area: Bergen
    monitored_conditions:
      - index
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

hfurubotten committed:
Add nilu air quality sensor platform

fabaff commented on homeassistant/components/sensor/nilu_air_quality.py:
```diff
@@ -0,0 +1,162 @@
+                  'Latitude and longitude must exist together'): cv.longitude,
+    vol.Optional(CONF_AREAS):
+        vol.All(cv.ensure_list, [vol.In(AREAS)]),
+})
```
I would like to suggest to take a different approach for the configuration. From 

```yaml
sensor:
  - platform: nilu_air_quality
    areas:
      - Brumunddal
      - Bergen
```

to

```yaml
sensor:
  - platform: nilu_air_quality
    area: bergen
    monitored_conditions:
      - pm10
      - o3
      - ...
```

This would more align with the configuration of the `luftdaten` sensor which is providing similar data.

Also, creating a sensor for every condition would more align with the configuration through the frontend.

hfurubotten replied:
I'll take the challenge! Moving the PR to WIP and will start poking the `luftdaten` sensor and draw inspiration from there. 
Thanks.

hfurubotten replied:
I have changed the configuration to only have one area for each sensor configuration entry, and added possibility to configure which data from the stations should be mounted as an sensor. The sensor has gained a lot of capabilities from it, and hopefully give a more rich sensor platform for users. I have now ran it on my local instance for several days and works really well with new configurations. 
Hopefully you agree that it is improved to the better now. Thanks!

fabaff commented on homeassistant/components/sensor/nilu_air_quality.py:
```diff
@@ -0,0 +1,162 @@
+
+CONF_AREAS = 'areas'
+
+AREAS = [
```
Why not use [the data](https://github.com/hfurubotten/niluclient/blob/master/niluclient/__init__.py#L19) from the used module?

hfurubotten replied:
It was meant to limit the configuration input. As far as I know, this means that I have to import the module globally in the py file since the configuration setup is executed once the sensor is loaded, and that wouldn't work well and as far as I know is discouraged in the HA developers documentation. 
If there is a way to fix this I will gladly implement re-use of the module. In the mean time I have changed the name of the const list to reflect proper use of it in the configuration.

fabaff commented on homeassistant/components/sensor/nilu_air_quality.py:
```diff
@@ -0,0 +1,162 @@
+    'Bergen'
+]
+
+POLLUTION_INDEX = [
```
Dito

hfurubotten committed:
Updated configuration possibilies.
- Can be shown on the map.
- Configurable sensor name.
- More conditions to monitor is added.
- Can find stations in an area, or get a specified station.

hfurubotten commented on homeassistant/components/sensor/nilu_air_quality.py:
```diff
@@ -0,0 +1,162 @@
+    def __init__(self, api_data):
+        """Initialize the sensor."""
+        self._api_data = api_data
+        self._name = "NILU " + api_data.data.area + " " + api_data.data.name
```
Have added the possibility to configure parts of the name in the new configuration options. Thanks!

hfurubotten commented on homeassistant/components/sensor/nilu_air_quality.py:
```diff
@@ -0,0 +1,162 @@
+        """Initialize the sensor."""
+        self._api_data = api_data
+        self._name = "NILU " + api_data.data.area + " " + api_data.data.name
+        self._site_data = None
```
Removed. Forgot to remove it after fetching the data directly in update method. Thanks!

hfurubotten commented on homeassistant/components/sensor/nilu_air_quality.py:
```diff
@@ -0,0 +1,162 @@
+    def update(self) -> None:
+        """Update the sensor."""
+        self._api_data.update()
+        self._attrs['updated'] = dt_util.now()
```
Removed. Didn't know about about `state.last_updated` as I have only seen when the state last changed in the UI.

hfurubotten commented on homeassistant/components/sensor/nilu_air_quality.py:
```diff
@@ -0,0 +1,162 @@
+
+        for sensor in sensors:
+            attr_name = sensor.component
+            self._attrs[attr_name + "_value"] = sensor.value
```
Fixed to format only 2 decimal places for the values added to state and attributes.

MartinHjelmare commented on homeassistant/components/sensor/nilu_air_quality.py:
```diff
@@ -0,0 +1,238 @@
+    name = config.get(CONF_NAME)
+    area = config.get(CONF_AREA)
+    station = config.get(CONF_STATION)
+    monitored_conditions = config.get(CONF_MONITORED_CONDITIONS)
```
Use `dict[key]` for required config keys and keys with default config schema values.

MartinHjelmare commented on homeassistant/components/sensor/nilu_air_quality.py:
```diff
@@ -0,0 +1,238 @@
+                  'Can only configure one specific station or '
+                  'stations in a specific area pr sensor. '
+                  'Please only configure station or area.'):
+        cv.string,
```
Why not allow a list of stations?

hfurubotten replied:
Thanks for the code review! 
I added it as a string to align with how the `area` input works, which is the same type of input. Originally I had `areas` as a list, but it was suggested in code review that it might work better if it was one sensor for each config entry. I transferred this thinking to the `station` input as well, to not have too many approaches to the config on one sensor platform.

MartinHjelmare replied:
It seemed natural to me to be able to specify a list of stations, since the `location_client` will return a list of stations and we already iterate a list of stations. But other than that I don't have an opinion.

hfurubotten replied:
The `cv.ensurelist` would support both our cases, so it wouldn't change much more than having an extra 's' on the end of the config name. I also can agree that its more natural within the code to have the list. 
If you guys think that it fits better to have a list there, I would gladly change it to a list 😃
Sorry for the delayed response, btw.

MartinHjelmare replied:
Sounds good with `cv.ensure_list`.

MartinHjelmare commented on homeassistant/components/sensor/nilu_air_quality.py:
```diff
@@ -0,0 +1,238 @@
+
+    for station in stations:
+        client = NiluData(nilu.create_station_client(station))
+        client.update()
```
Does this do I/O? We're in async context here so are not allowed to do regular I/O.

hfurubotten replied:
Yes, that one does I/O, so I have changed to sync. I'm used to having async as desired when doing I/O, so didn't come to mind that it might be wrong usage.

MartinHjelmare commented on homeassistant/components/sensor/nilu_air_quality.py:
```diff
@@ -0,0 +1,238 @@
+class NiluSensor(Entity):
+    """Single nilu station air sensor."""
+
+    ICON = 'mdi:cloud-outline'
```
Move this to the module level.

MartinHjelmare commented on homeassistant/components/sensor/nilu_air_quality.py:
```diff
@@ -0,0 +1,238 @@
+            self._state = nilu.POLLUTION_INDEX[max_index]
+
+            self._attrs[ATTR_MAX_POLLUTION_INDEX] = max_index
+            self._attrs[CONF_AREA] = self._api.data.area
```
Don't use `CONF_` constants for attribute keys, use `ATTR_`.

MartinHjelmare commented on homeassistant/components/sensor/nilu_air_quality.py:
```diff
@@ -0,0 +1,238 @@
+            for sensor in sensors:
+                attr_name = sensor.component
+                self._attrs[attr_name] = "{0:.2f}".format(sensor.value)
+                self._attrs[attr_name + "_" + CONF_UNIT_OF_MEASUREMENT] = \
```
Use new style string formatting instead of string concatenation.

MartinHjelmare replied:
And don't use `CONF_` constants for attribute keys.

hfurubotten committed:
Fixes from code review

balloob commented:
@fabaff should this integration be part of the new air pollutant component?

hfurubotten commented:
If you guys are planning to integrate air pollutants in a new component instead of having sensors, I would love to tag along by changing this sensor to fit within the new component. I'm guessing that I can take a look at [air_pollutants/demo.py](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/air_pollutants/demo.py) and base the configuration and output of data on that one. I can probably start the rewrite and test it locally next week. Just give me a little heads up and I'll begin! 😄 

(Getting all this information in a simple lovelace card for air pollution would do wonders for the presentation within many of my HA instances, but I'm guess that would be a bit more long term.)

fabaff commented:
Would make sense to migrate it to an `air_quality` platform.
